File: python/perfectSum.py
class Solution:
    def perfectSum(self, arr, target):
        n = len(arr)
        dp = [[0] * (target + 1) for _ in range(n + 1)]
        for i in range(n + 1):
            dp[i][0] = 1
        for i in range(1, n + 1):
            for j in range(target + 1):
                dp[i][j] = dp[i - 1][j]
                if arr[i - 1] <= j:
                    dp[i][j] += dp[i - 1][j - arr[i - 1]]
        return dp[n][target]


if __name__ == "__main__":
    obj = Solution()
    print(obj.perfectSum([1, 2, 3], 4))

File: src/com/gfg/dynamicProgramming/PerfectSum.java
package com.gfg.dynamicProgramming;

class PerfectSum {
    public static void main(String[] args) {
        Integer[] arr = { 1, 2, 3 };
        Integer target = 4;
        int result = getPerfectSum(arr, target);
        System.err.println(result);
    }

    public static Integer getPerfectSum(Integer[] arr, int target) {
        int n = arr.length;
        int[][] dp = new int[n + 1][target + 1];
        for (int i = 0; i <= n; i++) {
            dp[i][0] = 1;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < (target + 1); j++) {
                dp[i][j] = dp[i - 1][j];
                if (arr[i - 1] <= j) {
                    dp[i][j] += dp[i - 1][j - arr[i - 1]];
                }
            }
        }
        return dp[n][target];
    }
}